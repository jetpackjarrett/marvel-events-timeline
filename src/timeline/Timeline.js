import React, { Component } from 'react';
import { range } from 'lodash';
import { fetchFromMarvel } from '../services/marvelApi';
import EventDetail from '../eventDetail/EventDetail';
import TimelineEvent from './TimelineEvent';
import './timeline.css';

export default class Timeline extends Component {

  state = {
    eventsByYear: null,
    selectedEvent: null
  };

  componentDidMount() {
    fetchFromMarvel('events', { orderBy: 'startDate', limit: 100 })
      .then(this.setEvents);
  }

  setEvents = (response) => {
    const events = response.data.results.filter((result) => result.start && result.end);
    const years = range(new Date(events[0].start).getFullYear(), new Date().getFullYear());
    const eventsByYear = years.reduce((memo, year) => {
      memo[year] = events.filter(({ start }) => new Date(start).getFullYear() === year);
      return memo;
    }, {});
    this.setState({ eventsByYear });
  }

  selectEvent = (selectedEvent) => () => this.setState({ selectedEvent });

  deselectEvent = () => this.setState({ selectedEvent: null });

  render() {
    const { eventsByYear, selectedEvent } = this.state;

    if (!eventsByYear) {
      return <div className="loader" />;
    }

    return (
      <div className="timeline">
        {eventsByYear && Object.keys(eventsByYear).map((year) => {
          const events = eventsByYear[year];

          return (
            <section key={year} className="timeline-year">
              <header>
                <h2>{year}</h2>
              </header>
              <article className="timeline-year-events">
                {events.map((event) => (
                  <TimelineEvent
                    key={event.id}
                    handleClick={this.selectEvent(event)}
                    {...event} />
                ))}
              </article>
            </section>
          );
        })}
        {selectedEvent && <EventDetail {...selectedEvent} onClose={this.deselectEvent} />}
      </div>
    );
  }
}
