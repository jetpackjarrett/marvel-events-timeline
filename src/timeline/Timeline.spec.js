/*eslint import/no-extraneous-dependencies: ["error", { "devDependencies": true }] */
import React from 'react';
import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Timeline from './Timeline';

const mockResponse = {
  data: {
    results: [
      {
        start: '1995-10-21',
        end: '1995-10-28',
        id: 123,
        handleClick: jest.fn(),
        thumbnail: {
          extension: '.jpg',
          path: 'http://mock-image.com/fake',
        },
        title: 'Amazing Javascript #1',
      }
    ]
  }
};

jest.mock('../services/marvelApi', () => {
  return {
    getImageUrl: jest.fn(),
    fetchFromMarvel: jest.fn(() => Promise.resolve(mockResponse))
  };
});

describe('Timeline', () => {
  it('renders the timeline using data set to state from api fetch', () => {
    const component = mount(<Timeline />);
    component.instance().setEvents(mockResponse);
    expect(toJson(component)).toMatchSnapshot();
  });

  describe('selectEvent', () => {
    it('sets selectedEvent state to the given event', () => {
      const mockEvent = 'NOT_A_REAL_EVENT';
      const component = shallow(<Timeline />);
      component.instance().selectEvent(mockEvent)();
      expect(component.state().selectedEvent).toEqual(mockEvent);
    });
  });

  describe('deselectEvent', () => {
    it('sets selectedEvent state to null', () => {
      const component = shallow(<Timeline />);
      component.instance().deselectEvent();
      expect(component.state().selectedEvent).toEqual(null);
    });
  });
});
