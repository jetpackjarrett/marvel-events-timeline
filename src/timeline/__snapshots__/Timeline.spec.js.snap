exports[`Timeline renders the timeline using data set to state from api fetch 1`] = `
<Timeline>
  <div
    className="timeline">
    <section
      className="timeline-year">
      <header>
        <h2>
          1995
        </h2>
      </header>
      <article
        className="timeline-year-events">
        <TimelineEvent
          end="1995-10-28"
          handleClick={[Function]}
          id={123}
          start="1995-10-21"
          thumbnail={
            Object {
              "extension": ".jpg",
              "path": "http://mock-image.com/fake",
            }
          }
          title="Amazing Javascript #1">
          <div
            className="timeline-event"
            onClick={[Function]}
            style={
              Object {
                "backgroundImage": "url(null)",
              }
            }>
            <aside>
              <h3>
                Amazing Javascript #1
              </h3>
              <p>
                <time>
                  October 1995
                </time>
              </p>
            </aside>
          </div>
        </TimelineEvent>
      </article>
    </section>
    <section
      className="timeline-year">
      <header>
        <h2>
          1996
        </h2>
      </header>
      <article
        className="timeline-year-events" />
    </section>
    <section
      className="timeline-year">
      <header>
        <h2>
          1997
        </h2>
      </header>
      <article
        className="timeline-year-events" />
    </section>
    <section
      className="timeline-year">
      <header>
        <h2>
          1998
        </h2>
      </header>
      <article
        className="timeline-year-events" />
    </section>
    <section
      className="timeline-year">
      <header>
        <h2>
          1999
        </h2>
      </header>
      <article
        className="timeline-year-events" />
    </section>
    <section
      className="timeline-year">
      <header>
        <h2>
          2000
        </h2>
      </header>
      <article
        className="timeline-year-events" />
    </section>
    <section
      className="timeline-year">
      <header>
        <h2>
          2001
        </h2>
      </header>
      <article
        className="timeline-year-events" />
    </section>
    <section
      className="timeline-year">
      <header>
        <h2>
          2002
        </h2>
      </header>
      <article
        className="timeline-year-events" />
    </section>
    <section
      className="timeline-year">
      <header>
        <h2>
          2003
        </h2>
      </header>
      <article
        className="timeline-year-events" />
    </section>
    <section
      className="timeline-year">
      <header>
        <h2>
          2004
        </h2>
      </header>
      <article
        className="timeline-year-events" />
    </section>
    <section
      className="timeline-year">
      <header>
        <h2>
          2005
        </h2>
      </header>
      <article
        className="timeline-year-events" />
    </section>
    <section
      className="timeline-year">
      <header>
        <h2>
          2006
        </h2>
      </header>
      <article
        className="timeline-year-events" />
    </section>
    <section
      className="timeline-year">
      <header>
        <h2>
          2007
        </h2>
      </header>
      <article
        className="timeline-year-events" />
    </section>
    <section
      className="timeline-year">
      <header>
        <h2>
          2008
        </h2>
      </header>
      <article
        className="timeline-year-events" />
    </section>
    <section
      className="timeline-year">
      <header>
        <h2>
          2009
        </h2>
      </header>
      <article
        className="timeline-year-events" />
    </section>
    <section
      className="timeline-year">
      <header>
        <h2>
          2010
        </h2>
      </header>
      <article
        className="timeline-year-events" />
    </section>
    <section
      className="timeline-year">
      <header>
        <h2>
          2011
        </h2>
      </header>
      <article
        className="timeline-year-events" />
    </section>
    <section
      className="timeline-year">
      <header>
        <h2>
          2012
        </h2>
      </header>
      <article
        className="timeline-year-events" />
    </section>
    <section
      className="timeline-year">
      <header>
        <h2>
          2013
        </h2>
      </header>
      <article
        className="timeline-year-events" />
    </section>
    <section
      className="timeline-year">
      <header>
        <h2>
          2014
        </h2>
      </header>
      <article
        className="timeline-year-events" />
    </section>
    <section
      className="timeline-year">
      <header>
        <h2>
          2015
        </h2>
      </header>
      <article
        className="timeline-year-events" />
    </section>
    <section
      className="timeline-year">
      <header>
        <h2>
          2016
        </h2>
      </header>
      <article
        className="timeline-year-events" />
    </section>
  </div>
</Timeline>
`;
